browser/notifications/ads_notification_handler.cc

```diff
@@ -0,0 +1,67 @@
+/* Copyright (c) 2019 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "brave/browser/notifications/ads_notification_handler.h"
+
+#include <memory>
+
+#include "brave/browser/brave_ads/ads_service_factory.h"
+#include "brave/components/brave_ads/browser/ads_service.h"
+#include "url/gurl.h"
+
+namespace brave_ads {
+
+const void* const kUserDataKey = &kUserDataKey;
+
+AdsNotificationHandler::AdsNotificationHandler(Profile* profile)
+    : profile_(profile) {
+  DCHECK(profile_);
+}
+
+AdsNotificationHandler::~AdsNotificationHandler() = default;
+
+void AdsNotificationHandler::OnShow(Profile* profile, const std::string& id) {
+  AdsService* ads_service = AdsServiceFactory::GetForProfile(profile);
+  DCHECK(ads_service);
+
+  ads_service->OnShowAdNotification(id);
+}
+
+void AdsNotificationHandler::OnClose(Profile* profile,
+                                     const GURL& origin,
+                                     const std::string& id,
+                                     bool by_user,
+                                     base::OnceClosure completed_closure) {
+  AdsService* ads_service = AdsServiceFactory::GetForProfile(profile);
+  DCHECK(ads_service);
```

Comment:
are you sure it always exists? For instance, imagine we are closing a window and destroying the profile - could we have a race condition (i.e. all keyed services are destroyed before this call)?

Reply:
Added defensive code, thanks